staging/vhost-device-gpu/src/main.rs

```diff
@@ -10,52 +10,138 @@
 // rutabaga library is available for musl.
 #[cfg(target_env = "gnu")]
 pub mod gnu_main {
-    use std::{path::PathBuf, process::exit};
+    use std::{
+        path::{Path, PathBuf},
+        process::exit,
+    };
 
-    use clap::Parser;
+    use clap::{ArgAction, Parser, ValueEnum};
     use log::{error, info};
     use thiserror::Error as ThisError;
     use vhost_device_gpu::{
-        device::{self, VhostUserGpuBackend},
-        GpuConfig, GpuMode,
+        device, device::VhostUserGpuBackend, GpuCapset, GpuConfig, GpuFlags, GpuMode,
     };
     use vhost_user_backend::VhostUserDaemon;
     use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
 
-    type Result<T> = std::result::Result<T, Error>;
+    #[derive(ValueEnum, Debug, Copy, Clone, Eq, PartialEq)]
+    #[repr(u64)]
+    pub enum CapsetName {
+        /// [virglrenderer] OpenGL implementation, superseded by Virgl2
+        Virgl = GpuCapset::VIRGL.bits(),
+
+        /// [virglrenderer] OpenGL implementation
+        Virgl2 = GpuCapset::VIRGL2.bits(),
+
+        /// [gfxstream] Vulkan implementation (partial support only){n}
+        /// NOTE: Can only be used for 2D display output for now, there is no
+        /// hardware acceleration yet
+        #[cfg(feature = "gfxstream")]
+        GfxstreamVulkan = GpuCapset::GFXSTREAM_VULKAN.bits(),
+
+        /// [gfxstream] OpenGL ES implementation (partial support only){n}
+        /// NOTE: Can only be used for 2D display output for now, there is no
+        /// hardware acceleration yet
+        #[cfg(feature = "gfxstream")]
+        GfxstreamGles = GpuCapset::GFXSTREAM_GLES.bits(),
+    }
 
-    #[derive(Debug, ThisError)]
-    pub enum Error {
-        #[error("Could not create backend: {0}")]
-        CouldNotCreateBackend(device::Error),
-        #[error("Could not create daemon: {0}")]
-        CouldNotCreateDaemon(vhost_user_backend::Error),
-        #[error("Fatal error: {0}")]
-        ServeFailed(vhost_user_backend::Error),
+    impl From<CapsetName> for GpuCapset {
+        fn from(capset_name: CapsetName) -> GpuCapset {
+            GpuCapset::from_bits(capset_name as u64)
+                .expect("Internal error: CapsetName enum is incorrectly defined")
```

Comment:
it may be better to use `unwrap_or_else` to provide a fallback value like `GpuCapset::empty()`  and log the error instead of `expect`

Reply:
if the test guarantees that then it's fine as is.